src/manager.S

```diff
@@ -0,0 +1,165 @@
+.data
+
+comps_cnt:     .dword 0
+comps_addr:    .dword 0
+switcher_caps: .dword 0
+
+.text
+.balign 4
+
+#include "comps_offsets.h"
+
+// Variables
+.global comps_cnt
+.global comps_addr
+.global switcher_caps
+
+// Functions
+.global asm_call_wrapper
+.global init_compartments
+.global add_compartment
+.global del_compartment
+
+/**
+ * Wrapper to store/restore state when coming from C
+ *
+ * @param x0 ASM function to call
+ * @param x1-x6 parameters to pass to function in c0
+ */
+.type asm_call_wrapper, "function"
+asm_call_wrapper:
+    // Save `x0` so we can temporarily use it
+    cvtp      c0, x0
+    str       c0, [sp]
+
+    // Derive `clr` (in case asm function does something weird with `PCC`)
+    cvtp      c0, lr
+    swp       c0, c0, [sp]
+    sub       sp, sp, #16
+
+    blr       c0
+    ldr       clr, [sp, #16]!
+    ret       clr
+
+/**
+ * Sets up memory for compartments
+ *
+ * Initializes required memory for compartments. This involves allocating a
+ * memory region for use by switcher code, to contain required capabilities,
+ * and deriving appropriate PCC/DDC values containing the executable switcher
+ * code, and the aforementioned memory region, respectively.
+ *
+ * @return Pointer to newly allocated memory region
+ */
+.type init_compartments, "function"
+init_compartments:
+    // Compute size of required memory, equivalent to `length` parameter of
+    // `mmap`
+    mov       x0, #COMP_SIZE
+    mov       x1, #MAX_COMP_COUNT
+    mov       x2, #32    // size of the 2 switcher capabilities
+    madd      x1, x0, x1, x2
```

Comment:
I think that's correct, but you can just write a static expression and do it in a single instruction:
```
mov  x1, #(32 + (COMPSIZE * MAX_COMP_COUNT))
```

It'll fail if it's too big for a single `mov`, though. In that case, you can use a [special immediate syntax](https://sourceware.org/binutils/docs-2.36/as.html#AArch64_002dRelocations):

```
#define ALLOC_SIZE  (32 + (COMPSIZE * MAX_COMP_COUNT))
movz x1, #:abs_g1:ALLOC_SIZE
movk x1, #:abs_g0_nc:ALLOC_SIZE
```